fgpyo/fasta/sequence_dictionary.py

```diff
@@ -0,0 +1,441 @@
+
+    @property
+    def assembly(self) -> Optional[str]:
+        return self.get(Keys.ASSEMBLY)
+
+    @property
+    def uri(self) -> Optional[str]:
+        return self.get(Keys.URI)
+
+    @property
+    def species(self) -> Optional[str]:
+        return self.get(Keys.SPECIES)
+
+    @property
+    def description(self) -> Optional[str]:
+        return self.get(Keys.DESCRIPTION)
+
+    @property
+    def topology(self) -> Optional[Topology]:
+        value = self.get(Keys.TOPOLOGY)
+        if value is None:
+            return None
+        return Topology[value]
+
+    def same_as(self, other: "SequenceMetadata") -> bool:
+        """Returns true if the sequences share a common reference name (including aliases), have
+        the same length, and the same MD5 if both have MD5s."""
+        if self.length != other.length:
+            return False
+        elif self.name != other.name and other.name not in self.all_names:
+            return False
+        self_m5 = self.md5
+        other_m5 = other.md5
+        if self_m5 is None or other_m5 is None:
+            return True
+        else:
+            return self_m5 == other_m5
+
+    def to_sam(self) -> Dict[str, Any]:
+        """Converts the sequence metadata to a dictionary equivalent to one item in the
+        list of sequences from `pysam.AlignmentHeader#to_dict()`."""
+        meta_dict: Dict[str, Any] = {
+            f"{Keys.SEQUENCE_NAME}": self.name,
+            f"{Keys.SEQUENCE_LENGTH}": self.length,
+        }
+        if len(self.attributes) > 0:
+            meta_dict = {**meta_dict, **self.attributes}
+
+        return meta_dict
+
+    @staticmethod
+    def from_sam(meta: Dict[Union[Keys, str], Any], index: int) -> "SequenceMetadata":
+        """Builds a `SequenceMetadata` from a dictionary.  The keys must include the sequence
+        name (`Keys.SEQUENCE_NAME`) and length (`Keys.SEQUENCE_LENGTH`).  All other keys from
+        `Keys` will be stored in the resulting attributes.
+
+        Args:
+            meta: the python dictionary with keys from `Keys`.  This is typically the dictionary
+                  stored in the `"SQ"` level of the two-level dictionary returned by the
+                  `pysam.AlignmentHeader#to_dict()` method.
+            index: the 0-based index to use for this sequence
+        """
+        name = meta[Keys.SEQUENCE_NAME]
+        length = meta[Keys.SEQUENCE_LENGTH]
+        attributes = copy.deepcopy(meta)
+        del attributes[Keys.SEQUENCE_NAME]
+        del attributes[Keys.SEQUENCE_LENGTH]
+        return SequenceMetadata(name=name, length=length, index=index, attributes=attributes)
+
+    def __getitem__(self, key: str) -> str:
+        return self.attributes[key]
+
+    def __setitem__(self, key: str, value: str) -> None:
+        self.attributes[key] = value
+
+    def __delitem__(self, key: str) -> None:
+        del self.attributes[key]
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self.attributes)
```

Comment:
I feel like these should probably be augmented to also handle `SN` and `LN`.
For get/set it should route access to the properties.
For del I assume it should fail.
for iter I think it should probably chain iterators?

Reply:
Given I'd like to port over fgbio faithfullly, this would be breaking behavior [relative to fgbio](https://github.com/fulcrumgenomics/fgbio/blob/fc8e1c6a442a7609e7727c46b3fc377e0de05c41/src/main/scala/com/fulcrumgenomics/fasta/SequenceDictionary.scala#L159).  Skipping for now.

Reply:
I think this is the only sticking point - on reflection I'd say the fgbio impl is wrong.  Without reference to _implementation_ why would you expect iteration and other methods to _only_ work for optional attributes?  Including well-defined ones like `MD`, but not `LN`?

Reply:
I'll fix it here and in fgprimer so they both agree.